Split up RunTests

nilsvu opened the pull request:
## Proposed changes

Also run most tests without Charm++. Only run in a Charm++ environment if needed.

I find significant reductions in unit test runtime with this PR (could be the reduced overhead from eliminating Charm++ and/or the smaller executable). The whole test suit runs more than twice as fast on CI, depending on the configuration, cutting down 10-20 minutes from some CI builds.

| CI Build            | Old             | New             |   Speedup |
|------------------|-----------------|-----------------|-----------|
| gcc-9 Debug      | 12:26 | 4:30 |   2.8x |
| gcc-9 Release    | 8:28 | 2:15 |   3.8x |
| gcc-10 Debug     | 16:19 | 4:16 |   3.8x |
| gcc-10 Release   | 10:28 | 1:51 |   5.7x |
| gcc-11 Debug     | 12:25 | 4:36 |   2.7x |
| gcc-11 Release   | 7:05 | 1:51 |   3.8x |
| clang-10 Release | 9:13 | 2:19 |   4.0x |
| clang-11 Debug   | 32:16 | 8:09 |   4.0x |
| clang-11 Release | 2:21 | 1:47 |   1.3x |
| clang-13 Debug   | 21:21 | 3:45 |   5.7x |

(compared to https://github.com/sxs-collaboration/spectre/actions/runs/5835823727/job/15828143351)

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
To run unit tests you don't need to compile the giant `RunTests` executable anymore (or edit `RunSingleTest`). Just build the individual test executable (like `Test_LinearOperators`) and run it (like `./bin/Test_LinearOperators`). See the [Catch2 docs](https://github.com/catchorg/Catch2/blob/devel/docs/command-line.md) for details on running the test executables. You can still build _all_ unit tests with the `unit-tests` target and run test cases with `ctest` (no change there).
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Split DgMessages tests into separate lib

We want to test these in a Charm++ environment

nilsvu committed:
Fix some more test linking

nilsvu committed:
Install Catch2 with Homebrew on macOS CI

Spack 0.18 has an old version that throws errors which I haven't
investigated further.

nilsdeppe commented on docs/DevGuide/BuildOptimization.md:
```diff
@@ -130,12 +130,12 @@ the compilation time or memory usage.
 
```
I'd like to keep it until I have actually used the split executable approach. I have no experience to agree that this won't be needed anymore. I expect so, but I'm not sure.

nilsvu replied:
Ok done

nilsdeppe commented on tests/Unit/TestMain.cpp:
```diff
@@ -0,0 +1,45 @@
+
+  const int result = Catch::Session().run(argc, argv);
+
+  // In the case where we run all the non-failure tests at once we must ensure
```
[optional] I think this comment can get updated as we remove more of the old-style error tests and use `CHECK_THROWS_WITH` instead. Probably something like:
```cpp
// Clean up the python environment only after all tests have finished running,
// since there could be multiple tests run in a single executable launch.
```

If you decide to do this, the comment is also in the `TestCharmMain.cpp` file

nilsdeppe reviewed (changes requested):
Overall LGTM, one request: I'd like to keep RunSingleTest for now so I can try the split executables and be completely sure we don't need `RunSingleTest` anymore. Another small comment/wording suggestion. Please squash immediately.

Thanks for doing this big cleanup! :D

nilsdeppe commented:
Hmm, the macOS CI seems to be extremely unhappy. Not sure what is causing that, though.

nilsvu committed:
Replace RunTests with one test exec per lib

Also run most tests without Charm++. Only run in a
Charm++ environment if needed.

nilsvu commented:
Fixed it (I think). Looks like on macOS we still have to make test helper libs static.

nilsvu commented:
Rebased on #5335

nilsdeppe commented:
Any idea about the segfault test failure on macOS? I can't tell what's going on

nilsvu commented:
Fixed the macOS build by running `Test_ErrorHandling` `WITH_CHARM` (like a couple other test execs that require a Charm++ environment).

nilsdeppe commented:
Thanks!

On Wed, Aug 16, 2023, 16:25 Nils Vu ***@***.***> wrote:

> Fixed the macOS build by running Test_ErrorHandling WITH_CHARM (like a
> couple other test execs that require a Charm++ environment).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sxs-collaboration/spectre/pull/5340#issuecomment-1681224863>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHHPFXHBREOKZV2QHWO5M3XVUUEXANCNFSM6AAAAAA3NMBZLY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
